### PR TITLE
Add tests exposing BatchNormalization bug(s)? when used in GANs.

### PR DESCRIPTION
**Pull request by mjdietzx**
Reproducing:

- https://github.com/fchollet/keras/issues/5644
- https://github.com/fchollet/keras/issues/5644

Both tests fail due to bug. They will pass if `batch_normalization` flag is set to `False` in `generator` and `discriminator` for their respective tests.

### CODE REVIEW

**Comment by fchollet** [comment]
Since this is a unit test, it should be reducible to something much shorter and much simpler. This looks like an integration test for GANs. Surely the issue you're trying to evidence has nothing to do with GANs themselves.

**Comment by buggi22** [comment]
In case it might help, here's a short script that isolates and reproduces the error, running with the TensorFlow backend:
https://gist.github.com/buggi22/753ef6e8bde23d97310e354a8450a2b7

**Comment by fchollet** [comment]
@buggi22 this is helpful. It looks like an update reuse issue when applying a model that contained a BN layer to new inputs. The issue will deal with:

- BN layer `call` method
- base `Layer` `add_update` method
- `Container` `call` method

Anyone wants to look into it?

**Comment by abhaikollara** [comment]
In case it helps, 
- The issue only occurs for Tensorflow
- For the code given by @buggi22 the error disappears if you fit `m2` before creating `m3`. See [here](https://gist.github.com/abhaikollara/3411296b5c575ac012308307a4d6be1b)

**Comment by fchollet** [comment]
This is helpful. In particular the second point sounds like it should be enough to figure out the cause. Unfortunately I still won't have time to look into it for the time being. Hopefully someone else can...

**Comment by fchollet** [comment]
Fixed; added unit test.
